runners/direct-java/src/main/java/org/apache/beam/runners/direct/MultiStepCombine.java

```diff
@@ -0,0 +1,402 @@
+          GlobalCombineFn fn = ((PerKey) application.getTransform()).getFn();
+          return fn instanceof CombineFn && isApplicable(application.getInputs(), fn);
+        }
+        return false;
+      }
+
+      private <InputT> boolean isApplicable(
+          Map<TupleTag<?>, PValue> inputs, GlobalCombineFn<InputT, ?, ?> fn) {
+        if (inputs.size() == 1) {
+          PCollection<InputT> input =
+              (PCollection<InputT>) Iterables.getOnlyElement(inputs.values());
+          WindowingStrategy<?, ?> windowingStrategy = input.getWindowingStrategy();
+          // Triggering with count based triggers is not appropriately handled here. Disabling
+          // most triggers is safe, though more broad than is technically required.
+          boolean triggerApplicable = DefaultTrigger.of().equals(windowingStrategy.getTrigger());
+          boolean accumulatorCoderAvailable;
+          try {
+            Coder<?> accumulatorCoder =
+                fn.getAccumulatorCoder(input.getPipeline().getCoderRegistry(), input.getCoder());
+            accumulatorCoderAvailable = accumulatorCoder != null;
+          } catch (CannotProvideCoderException e) {
+            throw new RuntimeException(
+                String.format(
+                    "Could not construct an accumulator %s for %s. Accumulator %s for a %s may be"
+                        + " null, but may not throw an exception",
+                    Coder.class.getSimpleName(),
+                    fn,
+                    Coder.class.getSimpleName(),
+                    Combine.class.getSimpleName()),
+                e);
+          }
+          return triggerApplicable && accumulatorCoderAvailable;
+        }
+        return false;
+      }
+    };
+  }
+
+  static class Factory<K, InputT, AccumT, OutputT>
+      extends SingleInputOutputOverrideFactory<
+            PCollection<KV<K, InputT>>, PCollection<KV<K, OutputT>>,
+            PerKey<K, InputT, OutputT>> {
+    public static PTransformOverrideFactory create() {
+      return new Factory<>();
+    }
+
+    private Factory() {}
+
+    @Override
+    public PTransformReplacement<PCollection<KV<K, InputT>>, PCollection<KV<K, OutputT>>>
+    getReplacementTransform(
+        AppliedPTransform<
+            PCollection<KV<K, InputT>>, PCollection<KV<K, OutputT>>,
+            PerKey<K, InputT, OutputT>>
+            transform) {
+      checkState(
+          transform.getTransform().getFn() instanceof CombineFn,
+          "%s.matcher() should only match %s instances using %s, got %s",
+          MultiStepCombine.class.getSimpleName(),
+          Combine.PerKey.class.getSimpleName(),
+          CombineFn.class.getSimpleName(),
+          transform.getTransform().getFn().getClass().getName());
+      @SuppressWarnings("unchecked")
+      CombineFn<InputT, AccumT, OutputT> fn =
+          (CombineFn<InputT, AccumT, OutputT>) transform.getTransform().getFn();
```

Comment:
Here, too, the override should be in terms of the bits pulled from `CombinePayload`. They can be cast to Java UDFs for now, but I think it is only needed at the last moment before you pass to `MultiStepCombine`.